include/os/linux/kernel/linux/simd_x86.h

```diff
@@ -145,28 +148,43 @@ kfpu_fini(void)
 {
 	int cpu;
 
-	for_each_possible_cpu(cpu) {
-		if (zfs_kfpu_fpregs[cpu] != NULL) {
-			kfree(zfs_kfpu_fpregs[cpu]);
+	if (zfs_kfpu_fpregs != NULL) {
+		for_each_possible_cpu(cpu) {
+			if (zfs_kfpu_fpregs[cpu] != NULL) {
+				kmem_cache_free(zfs_kfpu_state_cache,
+				    zfs_kfpu_fpregs[cpu]);
+			}
 		}
+		kfree(zfs_kfpu_fpregs);
 	}
 
-	kfree(zfs_kfpu_fpregs);
+	kmem_cache_destroy(zfs_kfpu_state_cache);
 }
 
 static inline int
 kfpu_init(void)
 {
 	int cpu;
 
+	/*
+	 * need to use kmem_cache_create to force alignment! kmalloc is not
+	 * (yet) guarantueed to align in configurations.
+	 */
+	zfs_kfpu_state_cache = kmem_cache_create("zfs_kfpu_state",
+	    sizeof (union fpregs_state), 64, NULL, NULL, NULL, NULL, NULL, 0);
```

Comment:
These arguments aren't right for the kernel's `kmem_cache_create()` prototype.  Let's do this to force a minimum 64 byte alignment but I think it would also be beneficial to force it to be cache aligned SLAB_HWCACHE_ALIGN.

```
        zfs_kfpu_state_cache = kmem_cache_create("zfs_kfpu_state",
           sizeof (union fpregs_state), 64, SLAB_HWCACHE_ALIGN, NULL);
```

Reply:
see above - I also thought about SLAB_HWCACHE_ALIGN, but the spl wrapper does not allow passing in arbitray slabflags.